target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8068-ap3935.dts

```diff
@@ -295,55 +310,51 @@
 	};
 };
 
-&soc {
-	mdio1: mdio {
-		compatible = "virtual,mdio-gpio";
-		#address-cells = <1>;
-		#size-cells = <0>;
-
-		status = "okay";
+&mdio0 {
+	status = "okay";
 
-		pinctrl-0 = <&mdio0_pins>;
-		pinctrl-names = "default";
+	pinctrl-0 = <&mdio0_pins>;
+	pinctrl-names = "default";
 
-		gpios = <&qcom_pinmux 1 GPIO_ACTIVE_HIGH &qcom_pinmux 0 GPIO_ACTIVE_HIGH>;
+	gpios = <&qcom_pinmux 1 GPIO_ACTIVE_HIGH &qcom_pinmux 0 GPIO_ACTIVE_HIGH>;
 
-		phy1: ethernet-phy@1 {
-			reg = <1>;
-		};
+	phy1: ethernet-phy@1 {
+		reg = <1>;
+	};
 
-		phy2: ethernet-phy@2 {
-			reg = <2>;
-		};
+	phy2: ethernet-phy@2 {
+		reg = <2>;
 	};
 };
 
 &gmac0 {
 	status = "okay";
 
 	qcom,id = <0>;
-	mdiobus = <&mdio1>;
+	mdiobus = <&mdio0>;
+
+	pinctrl-0 = <&rgmii0_pins>;
+	pinctrl-names = "default";
 
-	phy-mode = "rgmii";
+	phy-mode = "rgmii-id";
```

Comment:
Can we put this in another pr or just later when the patch is merged upstream?

Reply:
Unfortunately I sorta sold the issue short in the PR description. Let me attempt to explain a bit better.

This is because, when netbooting, the boot rom pokes the right bits into the phy to put it into rgmii-id mode. Then, the `fixed-link` config in the current device tree keeps the phy from being initialized by the driver so the delay bits in the phy registers never get reset.

But if you power cycle the box and it boots from flash, then the LAN1 interface doesn't work (it links, it attempts to transmit frames, but they show up as errors on the uplinked switch, and it can't receive any frames). Removing the fixed-link config and wiring up the rgmii pins in this config causes the phy registers to be set by the driver, but by default the delays are disabled. 

Without the kernel patch, the MAC driver won't initialize because it doesn't recognize the `rgmii-id` phy mode. The phy driver supports this mode and will set the registers appropriately.

I'm happy to split the PR up further in any way you suggest, but the board doesn't work very well without this particular patch; this bit is kind of the point of the whole thing. I hope this explanation helps.

Reply:
Please split it... What about upstreaming? Do you want help? I can take care of sending this upstream!